File: symbol/symbol.go
// Package symbol 所有符号的权威双向查找表。禁止其他包 hardcode 符号字符串。
package symbol

// 显示用字符串常量
const (
	ArrowLeft  = " <- "
	ArrowRight = " -> "
	ArrowEq    = " = "
)

type Entry struct {
	Word       string   // 唯一英文名，如 "add"、"lbrace"
	Glyphs     []string // 该词对应的全部符号，如 {"+"}、{"!=","≠"}
	Precedence int      // Pratt 优先级，0=非中缀
	Arith, Cmp bool
	Unary      bool
}

// byWord 英文名 → 符号定义。Word 是唯一主键。
var byWord = map[string]Entry{
	// ── 分组符 ────────────────────────────────────────────────────
	"lparen":    {"lparen",    []string{"("}, 0, false, false, false},
	"rparen":    {"rparen",    []string{")"}, 0, false, false, false},
	"lbrace":    {"lbrace",    []string{"{"}, 0, false, false, false},
	"rbrace":    {"rbrace",    []string{"}"}, 0, false, false, false},
	"lbrack":    {"lbrack",    []string{"["}, 0, false, false, false},
	"rbrack":    {"rbrack",    []string{"]"}, 0, false, false, false},
	"comma":     {"comma",     []string{","}, 0, false, false, false},
	"semicolon": {"semicolon", []string{";"}, 0, false, false, false},
	"dot":       {"dot",       []string{"."}, 0, false, false, false},
	"colon":     {"colon",     []string{":"}, 0, false, false, false},
	// ── 箭头 / 赋值（= 兼作 copy opcode）──────────────────────────
	"assign": {"assign", []string{"<-", "->", "="}, 0, false, false, false},
	// ── 算术 ──────────────────────────────────────────────────────
	"add":  {"add",  []string{"+"}, 50, true, false, true},
	"sub":  {"sub",  []string{"-"}, 50, true, false, true},
	"pointer": {"pointer", []string{"*"}, 0, false, false, false},
		"mul":  {"mul",  []string{"×"}, 60, true, false, false},
	"div":  {"div",  []string{"÷"}, 60, true, false, false},
	"mod":  {"mod",  []string{"%"}, 60, true, false, false},
	// ── 比较 ──────────────────────────────────────────────────────
	"eq":  {"eq",  []string{"=="}, 30, false, true, false},
	"neq": {"neq", []string{"!=", "≠"}, 30, false, true, false},
	"lt":  {"lt",  []string{"<"}, 40, false, true, false},
	"gt":  {"gt",  []string{">"}, 40, false, true, false},
	"le":  {"le",  []string{"<=", "≤"}, 40, false, true, false},
	"ge":  {"ge",  []string{">=", "≥"}, 40, false, true, false},
	// ── 逻辑 ──────────────────────────────────────────────────────
	"and": {"and", []string{"&&"}, 20, false, true, false},
	"or":  {"or",  []string{"||"}, 10, false, true, false},
	"not": {"not", []string{"!"}, 0, false, true, true},
	// ── 数学符号 ──────────────────────────────────────────────────
	"math.sqrt": {"math.sqrt", []string{"√"}, 0, true, false, true},

	// ── 位运算 ────────────────────────────────────────────────────
	"bitand": {"bitand", []string{"&"}, 80, true, false, false},
	"bitor":  {"bitor",  []string{"|"}, 100, true, false, false},
	"bitxor": {"bitxor", []string{"^"}, 90, true, false, false},
	"shl":    {"shl",    []string{"<<"}, 70, true, false, false},
	"shr":    {"shr",    []string{">>"}, 70, true, false, false},
}

// byGlyph 符号 → Entry（init 自动构建）。
var byGlyph = map[string]*Entry{}

func init() {
	for i := range byWord {
		e := byWord[i]
		for _, g := range e.Glyphs {
			if prev, ok := byGlyph[g]; ok {
				panic("symbol: glyph " + g + " defined in both " + prev.Word + " and " + e.Word)
			}
			byGlyph[g] = &e
		}
	}
}

// ── 公开查询 ────────────────────────────────────────────────────

var zeroEntry Entry

func Lookup(glyph string) *Entry {
	if e := byGlyph[glyph]; e != nil { return e }
	return &zeroEntry
}
func ByWord(word string) *Entry {
	if e, ok := byWord[word]; ok { return &e }
	return &zeroEntry
}

func ScannerTwoCharOps() []string {
	var ops []string
	for _, e := range byWord {
		for _, g := range e.Glyphs {
			if len(g) == 2 && isASCII(g) { ops = append(ops, g) }
		}
	}
	return ops
}

func ScannerOneCharOps() []byte {
	var ops []byte
	for _, e := range byWord {
		for _, g := range e.Glyphs {
			if len(g) == 1 && g[0] < 128 { ops = append(ops, g[0]) }
		}
	}
	return ops
}

func UnicodeOps() []string {
	var ops []string
	for _, e := range byWord {
		for _, g := range e.Glyphs {
			if !isASCII(g) { ops = append(ops, g) }
		}
	}
	return ops
}

func isASCII(s string) bool {
	for i := 0; i < len(s); i++ { if s[i] >= 128 { return false } }
	return true
}

File: layout/src/symbol.rs
//! 所有符号的权威双向查找表（对齐 symbol/symbol.go）。禁止其他模块 hardcode 符号字符串。

// ── 显示用字符串常量 ─────────────────────────────────────────────────

pub const ARROW_LEFT: &str = " <- ";
pub const ARROW_RIGHT: &str = " -> ";
pub const ARROW_EQ: &str = " = ";

#[derive(Clone, Copy)]
pub struct Entry {
    pub word: &'static str,
    pub glyphs: &'static [&'static str],
    pub precedence: i32, // 0 = 非中缀
    pub arith: bool,
    pub cmp: bool,
    pub unary: bool,
}

impl Default for Entry {
    fn default() -> Self {
        Entry {
            word: "",
            glyphs: &[],
            precedence: 0,
            arith: false,
            cmp: false,
            unary: false,
        }
    }
}

// ── 符号表（word 唯一主键） ──────────────────────────────────────────

static ENTRIES: &[Entry] = &[
    // 分组符
    Entry {
        word: "lparen",
        glyphs: &["("],
        precedence: 0,
        arith: false,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "rparen",
        glyphs: &[")"],
        precedence: 0,
        arith: false,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "lbrace",
        glyphs: &["{"],
        precedence: 0,
        arith: false,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "rbrace",
        glyphs: &["}"],
        precedence: 0,
        arith: false,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "lbrack",
        glyphs: &["["],
        precedence: 0,
        arith: false,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "rbrack",
        glyphs: &["]"],
        precedence: 0,
        arith: false,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "comma",
        glyphs: &[","],
        precedence: 0,
        arith: false,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "semicolon",
        glyphs: &[";"],
        precedence: 0,
        arith: false,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "dot",
        glyphs: &["·"],
        precedence: 0,
        arith: false,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "colon",
        glyphs: &[":"],
        precedence: 0,
        arith: false,
        cmp: false,
        unary: false,
    },
    // 箭头 / 赋值（= 兼作 copy opcode）
    Entry {
        word: "assign",
        glyphs: &["<-", "->", "="],
        precedence: 0,
        arith: false,
        cmp: false,
        unary: false,
    },
    // 算术
    Entry {
        word: "add",
        glyphs: &["+"],
        precedence: 50,
        arith: true,
        cmp: false,
        unary: true,
    },
    Entry {
        word: "sub",
        glyphs: &["-"],
        precedence: 50,
        arith: true,
        cmp: false,
        unary: true,
    },
    Entry {
        word: "pointer",
        glyphs: &["*"],
        precedence: 0,
        arith: false,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "mul",
        glyphs: &["×"],
        precedence: 60,
        arith: true,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "div",
        glyphs: &["÷"],
        precedence: 60,
        arith: true,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "mod",
        glyphs: &["%"],
        precedence: 60,
        arith: true,
        cmp: false,
        unary: false,
    },
    // 比较
    Entry {
        word: "eq",
        glyphs: &["=="],
        precedence: 30,
        arith: false,
        cmp: true,
        unary: false,
    },
    Entry {
        word: "neq",
        glyphs: &["!=", "≠"],
        precedence: 30,
        arith: false,
        cmp: true,
        unary: false,
    },
    Entry {
        word: "lt",
        glyphs: &["<"],
        precedence: 40,
        arith: false,
        cmp: true,
        unary: false,
    },
    Entry {
        word: "gt",
        glyphs: &[">"],
        precedence: 40,
        arith: false,
        cmp: true,
        unary: false,
    },
    Entry {
        word: "le",
        glyphs: &["<=", "≤"],
        precedence: 40,
        arith: false,
        cmp: true,
        unary: false,
    },
    Entry {
        word: "ge",
        glyphs: &[">=", "≥"],
        precedence: 40,
        arith: false,
        cmp: true,
        unary: false,
    },
    // 逻辑
    Entry {
        word: "and",
        glyphs: &["&&"],
        precedence: 20,
        arith: false,
        cmp: true,
        unary: false,
    },
    Entry {
        word: "or",
        glyphs: &["||"],
        precedence: 10,
        arith: false,
        cmp: true,
        unary: false,
    },
    Entry {
        word: "not",
        glyphs: &["!"],
        precedence: 0,
        arith: false,
        cmp: true,
        unary: true,
    },
    // 数学符号
    Entry {
        word: "sqrt",
        glyphs: &["√"],
        precedence: 0,
        arith: true,
        cmp: false,
        unary: true,
    },
    // 位运算
    Entry {
        word: "bitand",
        glyphs: &["&"],
        precedence: 80,
        arith: true,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "bitor",
        glyphs: &["|"],
        precedence: 100,
        arith: true,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "bitxor",
        glyphs: &["^"],
        precedence: 90,
        arith: true,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "shl",
        glyphs: &["<<"],
        precedence: 70,
        arith: true,
        cmp: false,
        unary: false,
    },
    Entry {
        word: "shr",
        glyphs: &[">>"],
        precedence: 70,
        arith: true,
        cmp: false,
        unary: false,
    },
];

// ── 查询 ─────────────────────────────────────────────────────────────

/// 按 glyph 查 Entry（未命中返回零值）。
pub fn lookup(glyph: &str) -> Entry {
    for e in ENTRIES {
        if e.glyphs.contains(&glyph) {
            return *e;
        }
    }
    Entry::default()
}

/// 按 word 查 Entry（未命中返回零值）。
pub fn by_word(word: &str) -> Entry {
    for e in ENTRIES {
        if e.word == word {
            return *e;
        }
    }
    Entry::default()
}

fn is_ascii(s: &str) -> bool {
    s.bytes().all(|b| b < 128)
}

pub fn scanner_two_char_ops() -> Vec<String> {
    ENTRIES
        .iter()
        .flat_map(|e| e.glyphs.iter())
        .filter(|g| g.len() == 2 && is_ascii(g))
        .map(|g| g.to_string())
        .collect()
}

pub fn scanner_one_char_ops() -> Vec<u8> {
    ENTRIES
        .iter()
        .flat_map(|e| e.glyphs.iter())
        .filter(|g| g.len() == 1 && is_ascii(g))
        .map(|g| g.as_bytes()[0])
        .collect()
}

pub fn unicode_ops() -> Vec<String> {
    ENTRIES
        .iter()
        .flat_map(|e| e.glyphs.iter())
        .filter(|g| !is_ascii(g))
        .map(|g| g.to_string())
        .collect()
}
